1.rs

```diff
@@ -0,0 +1,59 @@
+use std::fs::File;
+use std::io::{self, BufRead};
+use std::path::Path;
+
+use std::collections::BinaryHeap;
+
+/* Each line contains the calories contained in an item in an elves' inventory.
+the inventories are separated by a blank line. 
+
+I need to print out the calorie total for the elf that has the most calories, and the sum of the top three elves.
+*/
+
+fn main() {
```

Comment:
(Disclaimer: This is mainly about rusty stuff, rather than the accuracy of the algorithm itself)

Could have main return a Result type, e.g.:
```
use std::result::Result;
...
fn main() -> Result<(), String> {
```

Then within the code, many of the `let Ok(x) = y` checks could be shortened to just `y?`. For example the current code could become:
```
    let lines = read_lines("./inputs/1.input").map_err(|err| err.to_string())?;
    let mut current_tally = 0;
    let mut heap = BinaryHeap::new();
    for line in lines {
        let calories = line.map_err(|err| err.to_string())?.parse::<i32>();
        if let Ok(result) = calories {
            current_tally += result;
        } else {
            ...
```

If we had wanted to fold the separate `read_lines` function into `main()` then that first line would become:
```
    let lines = io::BufReader::new(File::open("./inputs/1.input").map_err(|err| err.to_string())?).lines();
```

The `?` operator is a shorthand for "unpack the `Ok` result, or pass the `Err` result up the stack". In my example I've got the error type as a string, so I need to map any `io::Result` errors to a string before they can be returned via `main()` itself.

A good library for handling errors ergonomically is `anyhow`, which defines its own `anyhow::Result` type. Much of the above `map_err()` handling goes away since it has its own conversions built in: https://docs.rs/anyhow/latest/anyhow/

Reply:
Also one common catch with starting on rust is the difference between `str` and `String`. Can think of `str` as being a reference to a string buffer where the buffer is owned by someone else. In the case of my `Result<(), String>` example, it needs to be a full `String`, not just a `str` since we need to be able to pass ownership of the error message up the stack.

Reply:
Oh almost forgot that the coolest thing with using `anyhow` specifically is that you can build a sort of stack trace by adding on `context()` calls to errors. Sorta similar to go-style appending things to the `err` string, except with more structure.